clean up request and response version handling in managed handler

geoffkizer opened the pull request:
For request version: treat < 1.0 as an error, and > 1.1 should degrade to 1.1.
For response version: 1.x should return version as received; anything else should be treated as a bad response.

Add tests.

@stephentoub @davidsh @Priya91 @wfurt

geoffkizer commented:
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please
@dotnet-bot test Outerloop OSX x64 Debug Build please

stephentoub commented on src/System.Net.Http/src/Resources/Strings.resx:
```diff
@@ -378,6 +378,9 @@
     <value>HTTP 1.0 does not support chunking.</value>
   </data>
+  <data name="net_http_unsupported_version" xml:space="preserve">
+    <value>Request specified an unsupported HTTP version.</value>
```
Consider including in the error the requested version and also being prescriptive about what versions are allowed.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -505,29 +515,41 @@ private static bool LineIsEmpty(ArraySegment<byte> line)
                 line[4] != '/' ||
+                line[5] != '1' ||
+                line[6] != '.' ||
                 line[8] != ' ')
```
After this change, we should look at whether it's more efficient to simply reinterpret the first 8 bytes of the line as a long and then just do a comparison against an equivalent for the bytes for HTTP/1.0 and HTTP/1.1, e.g.
```C#
private static readonly ulong s_http10Bytes = BitConverter.ToUInt64(Encoding.ASCII.GetBytes("HTTP/1.0"), 0);
private static readonly ulong s_http11Bytes = BitConverter.ToUInt64(Encoding.ASCII.GetBytes("HTTP/1.1"), 0);
...
if (line.Length < 14 || // "HTTP/1.1 123\r\n" with optional phrase before the crlf
    line[8] != ' ')
{
    ThrowInvalidHttpResponse();
}

ulong first8Bytes = BitConverter.ToInt64(line.Slice(0, sizeof(long));
if (first8Bytes != s_http10Bytes && first8Bytes != s_http11Bytes)
{
   ThrowInvalidHttpResponse();
}
```

geoffkizer replied:
I'm reluctant to make to make this change without analyzing the perf, which I don't want to spend time on right now.  Good issue to investigate for the future though.

stephentoub replied:
Yup.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -505,29 +515,41 @@ private static bool LineIsEmpty(ArraySegment<byte> line)
-                !IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
+            // Set the response HttpVersion
+            byte minorVersion = line[7];
+            if (!IsDigit(minorVersion))
```
If we do my above suggestion, we can remove the digit check, as at that point we know it's already a 0 or 1.

If we don't do my suggestion, we can at least avoid doing the check until we know it's not a 0 or 1, and thus in 99.99% of the cases we'll not do the check, e.g.
```C#
byte minorVersion = line[7];
response.Version =
    minorVersion == '1' ? HttpVersion.Version11 ?
    minorVersion == '0' ? HttpVersion.Version10 ?
    !IsDigit(minorVersion) ? throw ... ?
    new Version(1, minorVersion - '0');
```

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -1949,6 +1949,12 @@ public static IEnumerable<object[]> VerifyUploadServersStreamsAndExpectedData
+            if (UseManagedHandler)
+            {
+                return;
+            }
```
It looks like .NET Framework also fails on 0,0?  It'd be good to validate that we throw with the same exception.

But for the purposes of this test, do we actually need to use 0,0?  The comment below suggests that we just need it to be something "unknown".  What happens if we instead use 1,9 or something like that?

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs:
```diff
@@ -14,6 +15,176 @@ public class HttpProtocolTests : HttpClientTestBase
+            });
+        }
+
+        [Fact]
```
These two tests are identical except for some data that could be parameterized.  How about making them a Theory instead?

stephentoub replied:
To help minimize the boilerplate, I'd suggest combining these two tests like this:
```C#
[SkipOnTargetFramework(TargetFrameworkMonikers.Uap)] // Uap does not support 1.0
[Fact]
public async Task GetAsync_RequestVersion10_Success() => GetAsync_RequestVersion_Success(new Version(1,0));

[Fact]
public async Task GetAsync_RequestVersion11_Success() => GetAsync_RequestVersion_Success(new Version(1,1));

private async Task GetAsync_RequestVersion_Success(Version version)
{
    ... // code of your test, setting request.Version to version, and using version.ToString(2) in the expected output
}
```

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs:
```diff
@@ -14,6 +15,176 @@ public class HttpProtocolTests : HttpClientTestBase
+        [InlineData(0, 9)]
+        public async Task GetAsync_BadVersion_Throws(int majorVersion, int minorVersion)
+        {
+            if (!UseManagedHandler)
```
Why are we making ManagedHandler different from every other handler we have?  It now differs from .NET Framework, WinHttpHandler, and CurlHandler?  That seems wrong.

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpProtocolTests.cs:
```diff
@@ -14,6 +15,176 @@ public class HttpProtocolTests : HttpClientTestBase
+        [InlineData(4, 2)]
+        public async Task GetAsync_UnknownVersion_DegradesTo11(int majorVersion, int minorVersion)
+        {
+            if (!UseManagedHandler)
```
Ditto

wfurt commented:
It looks ok to me. What is motivation for banning 0.9 requests? Is this what other handlers do?

geoffkizer commented:
@wfurt we don't support HTTP 0.9.  WinHttpHandler doesn't either, as far as I can tell.

geoffkizer commented:
I've updated the tests significantly to capture the behavior of different handlers.

This highlights a couple issues:

(1) We have a variety of different handlers with different behavior (managed handler, WinHttpHandler, CurlHandler, netfx, uap) and writing tests for all these cases is painful.  Suggestions on how to better organize the code to reflect the different handlers and their different behaviors is very welcome.
(2) The behavior is rather different across different handlers, as you can see in the conditional test code.  Some of these differences are arguably bugs, though most of them are not particularly high priority, given that they deal with obscure versioning issues.

The current behavior of the managed handler (as of this PR) is a combination of what the RFC specifies, combined with trying to match what seems to be the most sensible behavior of the existing handlers.  I'm not married to it, so if you think it's wrong, tell me.

I also addressed some review feedback from above.

geoffkizer commented:
@stephentoub @davidsh Opinions?  I'm happy to change behavior of certain cases if you want...

geoffkizer commented:
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please
@dotnet-bot test Outerloop OSX x64 Debug Build please
